### ISSUE TITLE
optax.softmax_cross_entropy_with_integer_labels Undexpected Behavior

### ISSUE THREAD

**Issue by EntilZha** [alpa-projects/alpa#860]
**Please describe the bug**

Through debugging my previous issue, we found that when using `optax.softmax_cross_entropy_with_integer_labels` instead of `optax.softmax_cross_entropy`, this would result in this warning:
```
/private/home/par/miniconda3/envs/jax/lib/python3.9/site-packages/alpa/shard_parallel/auto_sharding.py:870: UserWarning: Detect unexpected behaviors in the auto-sharding pass.
  warnings.warn("Detect unexpected behaviors in the auto-sharding pass.")
```
The model does seem to train, but unclear if there is something wrong going on silently.

The easiest way to reproduce this is in the alpa example at `examples/opt_finetune/run_clm_flax.py`, replace the block on loss like so:
```
# original
        #loss = optax.softmax_cross_entropy(
        #    shift_logits,
        #    jax.nn.one_hot(shift_labels, logits.shape[-1])
        # )
        loss = optax.softmax_cross_entropy_with_integer_labels(
            shift_logits,
            shift_labels,
        )
```

Our best guess is that the more efficient `with_integer_labels` version is using `jax.lax.stop_gradient`, and possible making something unhappy.
